src/macaron/repo_finder/provenance_extractor.py

```diff
@@ -0,0 +1,271 @@
+    if build_type == "https://slsa-framework.github.io/github-actions-buildtypes/workflow/v1":
+        repo = json_extract(build_def, ["externalParameters", "workflow", "repository"], str)
+
+    if not repo:
+        raise ProvenanceExtractionException("Failed to extract repository URL from provenance.")
+
+    # Extract the commit hash.
+    commit = ""
+    deps = json_extract(build_def, ["resolvedDependencies"], list)
+    for dep in deps:
+        if not isinstance(dep, dict):
+            continue
+        uri = json_extract(dep, ["uri"], str)
+        url = _clean_spdx(uri)
+        if url != repo:
+            continue
+        digest_set = json_extract(dep, ["digest"], dict)
+        commit = _extract_commit_from_digest_set(digest_set, intoto.v1.VALID_ALGORITHMS)
+
+    if not commit:
+        raise ProvenanceExtractionException("Failed to extract commit hash from provenance.")
+
+    return repo, commit
+
+
+def _extract_from_witness_provenance(payload: InTotoV01Payload) -> tuple[str, str]:
+    """Extract the repository and commit metadata from the witness provenance file found at the passed path.
+
+    To successfully return the commit and repository URL, the payload must respectively contain a Git attestation, and
+    either a GitHub or GitLab attestation.
+
+    Parameters
+    ----------
+    payload: InTotoPayload
+        The payload to extract from.
+
+    Returns
+    -------
+    tuple[str, str]
+        The repository URL and commit hash if found, a pair of empty strings otherwise.
+    """
+    predicate: dict[str, JsonType] | None = payload.statement.get("predicate")
+    if not predicate:
+        raise ProvenanceExtractionException("No predicate in payload statement.")
+
+    attestations = json_extract(predicate, ["attestations"], list)
+    commit = ""
+    repo = ""
+    for entry in attestations:
+        if not isinstance(entry, dict):
+            continue
+        entry_type = entry.get("type")
+        if not entry_type:
+            continue
+        if entry_type.startswith("https://witness.dev/attestations/git/"):
+            commit = json_extract(entry, ["attestation", "commithash"], str)
+        elif entry_type.startswith("https://witness.dev/attestations/gitlab/") or entry_type.startswith(
+            "https://witness.dev/attestations/github/"
+        ):
+            repo = json_extract(entry, ["attestation", "projecturl"], str)
+
+    if not commit or not repo:
+        raise ProvenanceExtractionException("Could not extract repo and commit from provenance.")
+
+    return repo, commit
+
+
+def _extract_commit_from_digest_set(digest_set: dict[str, JsonType], valid_algorithms: list[str]) -> str:
+    """Extract the commit from the passed DigestSet.
+
+    The DigestSet is an in-toto object that maps algorithm types to commit hashes (digests).
+    """
+    # TODO decide on a preference for which algorithm to accept.
```

Comment:
Git hashes are always SHA1 as far as I know, but to be vigilant I did a bit of digging through the `slsa-github-generator` code. If you look [here](https://github.com/slsa-framework/slsa-github-generator/blob/b595e069f78187eaead9bac15429f87237be550a/slsa/buildtype.go#L222), it clearly shows that the `.predicate.invocation.configSource.digest` field always contains a `"sha1"` key, which should be where we get the commit hash.

Reply:
Update: I found a case where the git hash is not provided through the "sha1" key: npm SLSA v1 provenances.
These provenances have the git hash in the "gitDigest" key ([npm provenance generator implementation](https://github.com/npm/provenance/blob/8a3ab1cbc231e72587cdbb5cd8113a592bfd401a/packages/generate-provenance/src/index.ts#L55C1-L55C43)).
I think it is probably best to assume it's either "sha1" or "gitDigest" unless we encounter any other exceptions (Have you encountered any?). I think this seems at least more precise than accepting a whole collection of hash algorithms.

Reply:
I will restrict it to `sha1` and `gitCommit` (for `in-toto` v1) and throw a detailed error for other values if they ever show up.